src/coreclr/vm/threads.h

```diff
@@ -997,14 +997,16 @@ class Thread
     // the top of the object.  Also, we want cache line filling to work for us
     // so the critical stuff is ordered based on frequency of use.
 
-    Volatile<ThreadState> m_State;   // Bits for the state of the thread
-
     // If TRUE, GC is scheduled cooperatively with this thread.
     // NOTE: This "byte" is actually a boolean - we don't allow
     // recursive disables.
     Volatile<ULONG>      m_fPreemptiveGCDisabled;
+    LONG                 m_generation;
```

Comment:
Can we fold `m_generation` and `m_fPreemptiveGCDisabled` together to avoid adding more instructions to PInvoke transitions? `m_fPreemptiveGCDisabled` is underutilized field - it is 32-bit, but we only use it as a bool.

Reply:
We flip m_fPreemptiveGCDisabled for various reasons, including when blocking a thread for suspension. 
If we reset generation at the same time, I think we will often reset generation too early. 

Like a thread was running a pinvoke and then tries to enter coop mode. It sets coop mode optimistically, checks the trap flag, then sees that we are suspending, so it becomes preemptive again and blocks itself. The generation should not be reset to 0 for this thread, since we did not run anything managed yet.

Reply:
To handle this case, can you snapshot the generation from m_fPreemptiveGCDisabled when starting the thread suspension, so that threads that wake up and block while the runtime is being suspended still get the benefit of the aging stacks?

Reply:
The current pattern for resetting generation on entering managed code is:
- set coop mode
- check if we need a slow path
- if do slow path, then then let the slow path deal with resetting generation, possibly after resuming.
- otherwise, reset generation and run managed code

Reply:
Yes, I understand that. I am wondering whether there is a way to do this without adding work to the PInvoke transitions. The PInvoke transitions in CoreCLR are more expensive than they need to be already.

Reply:
There may also be a concern with generation switching while we do GC. 
Like we see a thread blocked in preemptive and it is old, so we do not mark it. Then the thread wakes up and tries to enter coop mode, and blocks. If it resets age to 0, we could report its roots at relocation stage. It should be harmless, as the roots are too old, but it is one more scenario to think about.

Reply:
I need to think if a case when a thread goes out of preemptive mode concurrently with us doing suspension/GC is something we should worry about. 

If the actual target scenario is the threads that are blocked for more than one GC cycle, it might not matter if pinvoke resets generation together with disabling preemptive mode.  
Then we can just fold `m_fPreemptiveGCDisabled` and `m_generation` in on word.

Reply:
I just realized that this scheme would not work for native aot where we have `Thread::m_pTransitionFrame` that is a full pointer. We may want to put together design for the ideal most efficient PInvoke transition and eventually unify all runtimes on it.